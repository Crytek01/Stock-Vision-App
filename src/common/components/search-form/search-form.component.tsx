import { FunctionComponent, useEffect, useState } from "react";

import {
  Autocomplete,
  Button,
  Checkbox,
  FormControl,
  FormLabel,
  TextField,
} from "@mui/material";
import CheckBoxOutlineBlankIcon from "@mui/icons-material/CheckBoxOutlineBlank";
import CheckBoxIcon from "@mui/icons-material/CheckBox";

import { useDebouncedCallback } from "use-debounce";
import styles from "./search.module.css";

export interface ISearchFormOption {
  label: string;
  value: string;
}

interface ISearchFormProps {
  searchLabel: string;
  matchingFieldsLabel: string;
  addSelectedFieldsLabel: string;

  isNumericSearch?: boolean;

  onSelectedFields: (options: ISearchFormOption[]) => void;
  onSearchError?: () => void;

  fetcherProvider: (valueToSearch: string) => Promise<ISearchFormOption[]>;
}

const icon = <CheckBoxOutlineBlankIcon fontSize="small" />;
const checkedIcon = <CheckBoxIcon fontSize="small" />;

export const SearchForm: FunctionComponent<ISearchFormProps> = ({
  fetcherProvider,

  searchLabel,
  matchingFieldsLabel,
  addSelectedFieldsLabel,
  isNumericSearch = false,

  onSelectedFields,
  onSearchError,
}) => {
  const [searchQuery, setSearchQuery] = useState<string>("");
  const [options, setOptions] = useState<ISearchFormOption[]>([]);
  const [selectedOptions, setSelectedOptions] = useState<ISearchFormOption[]>(
    []
  );

  const searchAPIProviderDebounced = useDebouncedCallback(
    async (searchKey: string) => {
      try {
        if (!searchKey) {
          return setOptions([]);
        }

        const searchOptions = await fetcherProvider(searchKey);
        setOptions(searchOptions);
      } catch (error) {
        onSearchError?.();
      }
    },
    550
  );

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  const handleAddFieldsButton = () => {
    onSelectedFields(selectedOptions);
    setSearchQuery("");
    setSelectedOptions([]);
    setOptions([]);
  };

  const getOptionLabelAutoComplete = (option: ISearchFormOption) =>
    option.label;

  const isOptionEqualToValueAutoComplete = (
    option: ISearchFormOption,
    value: ISearchFormOption
  ) => {
    return option.label === value.label;
  };

  const handleSelectedOptionsChanged = (
    _: unknown,
    options: ISearchFormOption[]
  ) => {
    setSelectedOptions(options);
  };

  useEffect(() => {
    searchAPIProviderDebounced(searchQuery);
  }, [searchAPIProviderDebounced, searchQuery]);

  return (
    <div className={styles.container}>
      <div className={styles.searchContainer}>
        <FormControl fullWidth>
          <FormLabel>{searchLabel}</FormLabel>

          <TextField
            value={searchQuery}
            fullWidth
            onChange={handleSearchChange}
            type={isNumericSearch ? "number" : "text"}
          />
        </FormControl>
      </div>

      <div className={styles.matchingContainer}>
        <FormControl fullWidth>
          <FormLabel>{matchingFieldsLabel}</FormLabel>

          <Autocomplete
            value={selectedOptions}
            multiple
            options={options}
            disableCloseOnSelect
            getOptionLabel={getOptionLabelAutoComplete}
            isOptionEqualToValue={isOptionEqualToValueAutoComplete}
            onChange={handleSelectedOptionsChanged}
            renderOption={(props, option, { selected }) => (
              <li {...props}>
                <Checkbox
                  icon={icon}
                  checkedIcon={checkedIcon}
                  style={{ marginRight: 8 }}
                  checked={selected}
                />
                {option.label}
              </li>
            )}
            renderInput={(params) => <TextField {...params} />}
          />
        </FormControl>
      </div>

      <Button
        variant="contained"
        color="primary"
        onClick={handleAddFieldsButton}
        disabled={!selectedOptions.length}
      >
        {addSelectedFieldsLabel}
      </Button>
    </div>
  );
};
