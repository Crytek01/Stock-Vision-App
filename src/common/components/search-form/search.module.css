.container {
  display: flex;

  flex-direction: column;

  max-width: 30vh;

  padding: 1rem;
}

.searchContainer {
  margin-bottom: 1rem;
}

.matchingContainer {
  margin-bottom: 2rem;
}
