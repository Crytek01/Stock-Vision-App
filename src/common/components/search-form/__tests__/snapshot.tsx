import { render } from "@testing-library/react";
import { ISearchFormOption, SearchForm } from "../search-form.component";
it("renders search form component", () => {
  const { container } = render(
    <SearchForm
      searchLabel={""}
      matchingFieldsLabel={""}
      addSelectedFieldsLabel={""}
      onSelectedFields={function (options: ISearchFormOption[]): void {}}
      fetcherProvider={async function (
        _: string
      ): Promise<ISearchFormOption[]> {
        return [];
      }}
    />
  );
  expect(container).toMatchSnapshot();
});
