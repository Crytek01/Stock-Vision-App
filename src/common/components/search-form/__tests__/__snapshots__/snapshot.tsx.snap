// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders search form component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="searchContainer"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
        />
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="matchingContainer"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
        />
        <div
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                role="combobox"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    />
  </div>
</div>
`;
